File: app/src/main/java/com/ballastlane/android/texasam/data/repositories/user/RepositoryUser.java
package com.ballastlane.android.texasam.data.repositories.user;

import android.support.annotation.NonNull;

import com.ballastlane.android.texasam.data.entities.User;
import com.ballastlane.android.texasam.data.managers.remote.logon.LogonRequest;
import com.ballastlane.android.texasam.data.managers.remote.logon.LogonResponse;

import io.reactivex.Observable;
import io.reactivex.Single;
import okhttp3.ResponseBody;
import retrofit2.Response;

public class RepositoryUser implements RepositoryUserDataSource {

    private static RepositoryUser instance = null;
    private static RestUserSource sourceRest;

    @NonNull
    public static RepositoryUser newInstance() {
        if (instance == null) {
            instance = new RepositoryUser();
        }
        return instance;
    }


    private RestUserSource getRestSource() {
        if (sourceRest == null) {
            sourceRest = RestUserSource.newInstance();
        }
        return sourceRest;
    }


    @Override
    public Single<Response<ResponseBody>> checkEmail(String username) {
        return getRestSource().checkEmail(username);
    }

    @Override
    public Observable<LogonResponse> logon(LogonRequest logonRequest) {
        return getRestSource().logon(logonRequest);
    }

    @Override
    public Observable<User> get(String userId) {
        return getRestSource().get(userId);
    }


    @Override
    public Observable<User> getUser() {
        return getRestSource().getUser();
    }

//    @Override
//    public Observable<User> updateUser(UserRequest userRequest) {
//        return getRestSource().updateUser(userRequest);
//    }


}

File: sample/src/main/java/com/ballastlane/ballastlanebaseapp/data/repositories/user/RepositoryUser.kt
package com.ballastlane.ballastlanebaseapp.data.repositories.user

import com.ballastlane.ballastlanebaseapp.data.repositories.user.source.RestUserSource
import com.ballastlane.ballastlanebaseapp.entities.User

import io.reactivex.Single

/**
 * Created by Mariangela Salcedo (mariangelasalcedo@ballastlane.com) on 4/3/18.
 * Copyright (c) 2018 Ballast Lane Applications LLC. All rights reserved.
 */

class RepositoryUser : RepositoryUserDataSource {

    private val restSource: RestUserSource
        get() {
            if (sourceRest == null) {
                sourceRest = RestUserSource.newInstance()
            }
            return sourceRest as RestUserSource
        }

    override fun get(userId: String): Single<User> {
        return restSource.get(userId)
    }

    override fun create(id: String, name: String): Single<User> {
        return restSource.create(id, name)
    }

    override fun update(id: String, name: String): Single<User> {
        return restSource.update(id, name)
    }

    companion object {

        private var instance: RepositoryUser? = null
        private var sourceRest: RestUserSource? = null

        fun newInstance(): RepositoryUser {
            if (instance == null) {
                instance = RepositoryUser()
            }
            return instance as RepositoryUser
        }
    }

}
